### PR TITLE
tables: fix prefix index, when the charset is utf8, truncate it from runes

### LINKED ISSUE #7104: Prefix index implementation for utf8 string is incorrect
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```sql
mysql> show create table t1;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                         |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `name` varchar(12) DEFAULT NULL,
  KEY `pname` (`name`(12))
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> insert into t1 values('借款策略集_网页');
Query OK, 1 row affected (0.14 sec)
```

2. What did you expect to see?
```sql
mysql> select * from t1 where name = '借款策略集_网页';
+------------------------+
| name                   |
+------------------------+
| 借款策略集_网页        |
+------------------------+
1 row in set (0.00 sec)
```

3. What did you see instead?
```sql
mysql> select * from t1 where name = '借款策略集_网页';
Empty set (0.01 sec)

mysql> alter table t1 drop index pname;
Query OK, 0 rows affected (0.33 sec)

mysql> select * from t1 where name = '借款策略集_网页';
+------------------------+
| name                   |
+------------------------+
| 借款策略集_网页        |
+------------------------+
1 row in set (0.00 sec)

mysql> create index pname on t1 (name(12));
Query OK, 0 rows affected (0.34 sec)

mysql> select * from t1 where name = '借款策略集_网页';
Empty set (0.00 sec)

mysql> select * from t1 where name < '借款策略集_网页';
+--------------+
| name         |
+--------------+
| 借款策略     |
+--------------+
1 row in set (0.01 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
Release Version: v2.0.0
Git Commit Hash: 637e130e6a9ba2e54e158131c0466233db39a60e
Git Branch: HEAD
UTC Build Time: 2018-07-16 08:54:50
GoVersion: go version go1.10.1 darwin/amd64
TiKV Min Version: 2.0.0-rc.4.1
```

In our prefix index implementation, the index length is counted by its bytes. 

However, 

> The effective maximum length of a VARCHAR in MySQL 5.0.3 and later is subject to the maximum row size (65,535 bytes, which is shared among all columns) and the character set used. For example, utf8 characters can require up to three bytes per character, so a VARCHAR column that uses the utf8 character set can be declared to be a maximum of 21,844 characters.

### PR DESCRIPTION
**Pull request by winkyao**
## What have you changed? (mandatory)

fix #7104  , before this PR, the index length is counted by its bytes. when the charset is utf8 or utf8mb4, the length should be counted by its runes. This PR fixes this.

## What is the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

UT

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

NO

## Does this PR affect tidb-ansible update? (mandatory)

NO

## Does this PR need to be added to the release notes? (mandatory)

release note:

```
fix prefix index, when the charset is utf8 or utf8mb4, truncate it from runes.
```

### CODE REVIEW

File: table/tables/index.go [hunk 1]
```diff
@@ -124,6 +126,32 @@ func (c *index) getIndexKeyBuf(buf []byte, defaultCap int) []byte {
...
+		v := &indexedValues[i]
+		if v.Kind() == types.KindString || v.Kind() == types.KindBytes {
+			ic := c.idxInfo.Columns[i]
+			if ic.Tp.Charset == charset.CharsetUTF8 || ic.Tp.Charset == charset.CharsetUTF8MB4 {
+				val := v.GetBytes()
+				if ic.Length != types.UnspecifiedLength && utf8.RuneCount(val) > ic.Length {
+					rs := bytes.Runes(val)
+					truncateStr := string(rs[:ic.Length])
+					// truncate value and limit its length
+					v.SetBytes([]byte(truncateStr))
```
Comment (coocood) [inline table/tables/index.go]: `SetString` can save a memory allocation.

File: table/tables/index.go [hunk 2]
```diff
@@ -124,6 +126,32 @@ func (c *index) getIndexKeyBuf(buf []byte, defaultCap int) []byte {
...
 
+// truncateIndexValuesIfNeeded truncate the index values that be created that use only the leading part of column values.
+func (c *index) truncateIndexValuesIfNeeded(indexedValues []types.Datum) []types.Datum {
+	for i := 0; i < len(indexedValues); i++ {
+		v := &indexedValues[i]
+		if v.Kind() == types.KindString || v.Kind() == types.KindBytes {
+			ic := c.idxInfo.Columns[i]
+			if ic.Tp.Charset == charset.CharsetUTF8 || ic.Tp.Charset == charset.CharsetUTF8MB4 {
+				val := v.GetBytes()
+				if ic.Length != types.UnspecifiedLength && utf8.RuneCount(val) > ic.Length {
```
Comment (birdstorm) [inline table/tables/index.go]: I think we can use utf8.RuneCountInString(v) instead, and thus eliminate the usage of val.

Comment (winkyao) [reply table/tables/index.go]: But RuneCountInString need to convert bytes to string first, it's unworthy.

**Comment by winkyao** [comment]
@coocood @birdstorm PTAL

File: table/tables/index.go (hunk 2 above)
Comment (birdstorm) [reply table/tables/index.go]: I see.

**Comment by coocood** [comment]
LGTM

**Pull request merged**